pubsub/stanpubsub/stan.go

```diff
@@ -0,0 +1,538 @@
+		return nil, fmt.Errorf("open subscription %v: failed to open default connection: %v", u, err)
+	}
+	return opener.OpenSubscriptionURL(ctx, u)
+}
+
+func (o *defaultDialer) defaultConn() (*URLOpener, error) {
+	o.init.Do(o.createdOpener)
+	return o.opener, o.err
+}
+
+func (o *defaultDialer) createdOpener() {
+	serverURL := os.Getenv("STAN_SERVER_URL")
+	if serverURL == "" {
+		o.err = errors.New("STAN_SERVER_URL environment variable not set")
+		return
+	}
+	clusterID := os.Getenv("STAN_CLUSTER_ID")
+	if clusterID == "" {
+		o.err = errors.New("STAN_CLUSTER_ID environment variable not set")
+	}
+	clientID := os.Getenv("STAN_CLIENT_ID")
+	if clientID == "" {
+		o.err = errors.New("STAN_CLIENT_ID environment variable not set")
+	}
+	o.connect(serverURL, clusterID, clientID)
+}
+
+func (o *defaultDialer) connect(serverURL, clusterID, clientID string) {
+	conn, err := stan.Connect(clusterID, clientID, stan.NatsURL(serverURL))
+	if err != nil {
+		o.err = fmt.Errorf("failed to dial STAN_SERVER_URL %q: STAN_CLUSTER_ID: %q STAN_CLIENT_ID: %q,  %v", serverURL, clusterID, clientID, err)
+		return
+	}
+	o.opener = &URLOpener{Connection: conn}
+}
+
+// Scheme is the URL scheme stanpubsub registers its URLOpeners under on pubsub.DefaultMux.
+const Scheme = "stan"
+
+// URLOpener opens NATS URLs like "stan://mysubject".
+//
+// The URL host+path is used as the subject.
+//
+// No topic query parameters are supported.
+//
+// Following query parameters are supported for subscription:
+//	- queue 		- sets the name of the queue group.
+//	- durableName	- sets the durable name for the subscription.
+//	- maxInflight 	- sets the maximum number of messages the cluster will send without an ACK.
+//	- startSequence - sets the desired start sequence position and state
+//	- ackWait 		- sets the timeout for waiting for an ACK from the cluster's point of view for delivered messages
+type URLOpener struct {
+	// Connection to use for communication with the server.
+	Connection stan.Conn
+	// TopicOptions specifies the options to pass to OpenTopic.
+	TopicOptions TopicOptions
+	// SubscriptionOptions specifies the options to pass to OpenSubscription.
+	SubscriptionOptions []stan.SubscriptionOption
+}
+
+// OpenTopicURL opens a pubsub.Topic based on u.
+func (o *URLOpener) OpenTopicURL(ctx context.Context, u *url.URL) (*pubsub.Topic, error) {
+	for param := range u.Query() {
+		return nil, fmt.Errorf("open topic %v: invalid query parameter %s", u, param)
+	}
+	subject := path.Join(u.Host, u.Path)
+	return OpenTopic(o.Connection, subject, &o.TopicOptions)
+}
+
+// OpenSubscriptionURL opens a pubsub.Subscription based on input url.
+func (o *URLOpener) OpenSubscriptionURL(ctx context.Context, u *url.URL) (*pubsub.Subscription, error) {
+	opts := o.SubscriptionOptions
+	var queue string
+	for param, values := range u.Query() {
+		switch param {
+		case "queue":
+			queue = values[0]
+		case "durableName":
+			opts = append(opts, stan.DurableName(values[0]))
```

Comment:
What are the semantics if this is specified twice? I.e., I can create an `URLOpener` with `SubscriptionOptions` already set, and then call `OpenSubscriptionURL` on it with URL parameters set. The URL parameters should override the ones from the `URLOpener`.

The slice of `stan.SubscriptionOption` is neat, but it might be more clear to use a regular `SubscriptionOptions` struct with explicit fields, and then convert them into `stan` options at the end.

Reply:
The default `stan.SubscriptionOptions` has a field called `StartAt` which in my opinion is defined in a better way by using `stan.SubscriptionOption`, that's why I wanted to use the slice of these options.
But right now I've defined a new type `SubscriptionOptions` which tries to solve these two problems at once. 
It has methods which sets these option along with the subscription `StartPosition`, but also it has a `Validate` method which catches possible errors early.